Use full error messages for relations validation.

manwithtwowatches opened the pull request:
https://www.openproject.org/work_packages/13948

This PR should give sensible translated error messages when validating relations. Sadly they still appear at the top of the screen.

Also it should now respect the visibility of related work packages. So, as described in the ticket, if you make a relation and then the user loses the rights to view that work package then the relation will no long be visible (and there won't be an empty table row).

None of this is tested - sorry I've run out of time. Perhaps the reviewer @hschink could take over the responsibility;)

Commit:
Use full error messages for relations validation.

TeatroIO commented:
I've prepared a stage. <a href='http://fix-add-remove-relationships-validation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

mtakac2 commented on lib/api/errors/validation.rb:
```diff
@@ -41,7 +41,7 @@ def initialize(obj, args = { })
 
       def errors
-        @obj.errors.messages.map{ |m| { key: m[0], messages: m[1] }}
+        @obj.errors.full_messages
```
@manwithtwowatches :+1: I was just thinking about this. This basically means that we need to change it on the frontend. Is there some directive or service that deals with errors or do we need to change it manually for every resource that can be validated?

manwithtwowatches replied:
@mtakac error handling on the front end is pretty thin. api-helper i think is the only general solution but i did that recently and it's only being used in one place i think.

manwithtwowatches replied:
@mtakac actually thinking again there must be something that's broken. I will have a look through.

Commit:
Respecting visiblity of related work packages.

hschink commented on app/assets/javascripts/angular/helpers/components/work-packages-helper.js:
```diff
@@ -164,11 +164,19 @@ angular.module('openproject.workPackages.helpers')
+      var link = WorkPackagesHelper.getRelatedWorkPackageLink(workPackage, relation);
+      if(link) {
+        return link.fetch();
+      }
```
What happens if there is no link?

hschink commented on lib/api/v3/work_packages/relation_representer.rb:
```diff
@@ -55,11 +55,15 @@ def initialize(model, options = {}, *expand)
-          { href: "#{root_url}api/v3/work_packages/#{represented.model.to_id}" }
+          {
+            href: "#{root_url}api/v3/work_packages/#{represented.model.to_id}"
+          } if represented.model.to.visible?(@current_user)
```
This approach leads to rather complex code on the client side. I think that if a related work package is not visible it shouldn't even appear in the list of relations (otherwise the user gains knowledge about a work package she is not allow to see). Thus, instead, I propose to filter [the list of relations in the work package representer](https://github.com/opf/openproject/blob/dev/lib/api/v3/work_packages/work_package_representer.rb#L183-L185).

Commit:
Test that forbidden relations are not part of the WP model

Forbidden relations are relations to WPs that are not visible to the current user.

Commit:
Retrieve only relations with visible work packages

Commit:
Test parent link does not exist if parent is not visible

Commit:
Show parent link only if parent is visible to user

Commit:
Refactor specs

Commit:
Test children visibility is respected

Commit:
Respect child visibility

Commit:
Revert "Respecting visiblity of related work packages."

This reverts commit fb57024e5a49f2d57c6c27d12e16baccd7d633dd.

mtakac2 commented on lib/api/v3/work_packages/work_package_model.rb:
```diff
@@ -150,7 +150,9 @@ def watchers
         def relations
-          work_package.relations.map{ |relation| RelationModel.new(relation) }
+          relations = work_package.relations
+          visible_relations = relations.find_all { |relation| relation.other_work_package(work_package).visible? }
```
I think we should move this to the representer. Model should contaion only properties that can be set, representing the resource in different states should be the responsibility of representer.

ulferts replied:
But what keeps us from setting the relations as well in here. We can do it. And as long as we have the models, they should contain the business logic. And something like visibility is business logic, not representation logic. Guess again, something we need to discuss...

ulferts commented:
I am merging this now in order to close the bug. While there is some talk left to be done I don't think it justifies having the bug on OP.org.
